crates/compiler/mono/src/layout.rs

```diff
@@ -38,11 +38,397 @@ roc_error_macros::assert_sizeof_default!(Layout, 4 * 8);
 roc_error_macros::assert_sizeof_default!(UnionLayout, 3 * 8);
 roc_error_macros::assert_sizeof_default!(LambdaSet, 3 * 8);
 
+type LayoutResult<'a> = Result<Layout<'a>, LayoutProblem>;
+type RawFunctionLayoutResult<'a> = Result<RawFunctionLayout<'a>, LayoutProblem>;
+
+#[derive(Debug, Clone, Copy)]
+struct CacheMeta {
+    /// Does this cache entry include a recursive structure? If so, what's the recursion variable
+    /// of that structure?
+    has_recursive_structure: Option<Variable>,
+}
+
+impl CacheMeta {
+    #[inline(always)]
+    fn to_criteria(self) -> CacheCriteria {
+        let CacheMeta {
+            has_recursive_structure,
+        } = self;
+        CacheCriteria {
+            has_naked_recursion_pointer: false,
+            has_recursive_structure,
+        }
+    }
+}
+
+/// A single layer of the layout cache.
+/// Snapshots are implemented by operating on new layers, and rollbacks by dropping the latest
+/// layer.
+#[derive(Debug)]
+struct CacheLayer<Result>(MutMap<Variable, (Result, CacheMeta)>);
+
+impl<Result> Default for CacheLayer<Result> {
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
+
+#[cfg(debug_assertions)]
+#[derive(Debug, Default, Clone, Copy)]
+pub struct CacheStatistics {
+    pub hits: u64,
+    pub misses: u64,
+    /// How many times we could not cache a calculated layout
+    pub non_insertable: u64,
+    /// How many time we could not reuse a cached layout
+    pub non_reusable: u64,
+    /// How many times an entry was added to the cache
+    pub insertions: u64,
+}
+
+macro_rules! inc_stat {
+    ($stats:expr, $field:ident) => {
+        #[cfg(debug_assertions)]
+        {
+            $stats.$field += 1;
+        }
+    };
+}
+
+/// Layout cache to avoid recomputing [Layout] from a [Variable] multiple times.
+#[derive(Debug)]
+pub struct LayoutCache<'a> {
+    pub target_info: TargetInfo,
+    cache: std::vec::Vec<CacheLayer<LayoutResult<'a>>>,
+    raw_function_cache: std::vec::Vec<CacheLayer<RawFunctionLayoutResult<'a>>>,
+
+    /// Statistics on the usage of the layout cache.
+    #[cfg(debug_assertions)]
+    stats: CacheStatistics,
+    #[cfg(debug_assertions)]
+    raw_function_stats: CacheStatistics,
+
+    _marker: std::marker::PhantomData<&'a u8>,
```

Comment:
why is the marker needed? `'a` already occurs in some of the other types

Reply:
Good call, it's not needed anymore.